### ISSUE TITLE
Node freezes up after process restart

### ISSUE THREAD

**Issue by a-robinson** [cockroachdb/cockroach#9658]
This started from a problem @jseldess found while verifying https://github.com/cockroachdb/docs/pull/676. When you have cockroachdb running across 3 VMs in docker containers and tell one to quit, a fresh container comes up to take its place using the same storage. The problem is that SQL queries sent to it hang in the bowels of the storage and kv code.

In the past (https://github.com/cockroachdb/docs/pull/676#issuecomment-250778806), I've seen the main query-processing thread and a few others get stuck in `storage.(Replica).redirectOnOrAcquireLease`, but this afternoon's repro has a number of threads (including a query-processing thread) stuck in `kv.(*DistSender).sendToReplicas`.

I grabbed a bunch of debug info, but was interestingly unable to get info from the `/_status/raft` page because requests to that endpoint also hang.

Reproducing is really easy if you follow the instructions in https://github.com/cockroachdb/docs/pull/676. I haven't yet tried reproducing it on raw VMs rather than using containers.

@tschottdorf @tamird 

[stopper.txt](https://github.com/cockroachdb/docs/files/503757/stopper.txt)
[goroutines.txt](https://github.com/cockroachdb/docs/files/503758/goroutines.txt)
[gossip.txt](https://github.com/cockroachdb/docs/files/503756/gossip.txt)
[raft-status.txt](https://github.com/cockroachdb/docs/files/503755/raft-status.txt)
[ranges.txt](https://github.com/cockroachdb/docs/files/503754/ranges.txt)

**Comment by tschottdorf** [comment]
> but this afternoon's repro has a number of threads (including a query-processing thread) stuck in kv.(*DistSender).sendToReplicas.

That's just the "other side" of the goroutine waiting on Raft (which always has `addWriteCmd` in the stack trace; there are a few of those in your goroutines.txt above).

`ranges.txt` shows a lot of ranges which are quiescent/dormant. Maybe they shouldn't be. It would be interesting where the addWriteCommands are stuck. I'll see if I can repro this.

**Comment by tschottdorf** [comment]
Seeing a lot of this in the logs:

```
I161003 22:17:54.059260 1 kv/dist_sender.go:1039  [n?,range-lookup] sending to all 2 replicas failed; last error: rpc error: code = 14 desc = grpc: the connection is unavailable
W161003 22:17:54.059341 1 kv/dist_sender.go:666  [n?] failed to send RPC: sending to all 2 replicas failed; last error: rpc error: code = 14 desc = grpc: the connection is unavailable
```

In the current cluster, despite restarting every node, two out of three nodes are unable to even do a `SELECT 1`.

**Comment by a-robinson** [comment]
I got the cluster re-hung for you. `show databases` hangs, while `select 1` executes just fine.

**Comment by tschottdorf** [comment]
Somehow, the Raft transports can't make progress. All Replicas are stuck in candidate mode and continuously vote for themselves. From `docker logs $(docker ps -q) 2>&1 | grep 'raft transport stream'` I can guess that only the following Raft streams are active (these are ReplicaIDs of Range 1)

```
3 <--
|     \
v      \
1  <--- 2

```

This obviously isn't going to help anyone making progress. As to why it's stuck like that, looking at Gossip (still talking ReplicaID of Range 1).

We have the following containers:
```
1: 63259e48747c
2: b2bb4029cad9
3: d67a7b9fb60c
```

```
1 looks fully connected but is in fact stalled:
gossip status (stalled, 3 nodes)
gossip client (0/3 cur/max conns)
gossip server (2/3 cur/max conns, infos 38348/30519 sent/received, bytes 7469588B/5676056B sent/received)
  2: b2bb4029cad9:26257 (10h35m16s)
  3: d67a7b9fb60c:26257 (10h35m58s)
```

```
2 is also stalled:
gossip status (stalled, 3 nodes)
gossip client (2/3 cur/max conns)
  1: cockroachdb-0:26257 (10h45m0s: infos 15374/23181 sent/received, bytes 2854928B/4193231B sent/received)
  3: d67a7b9fb60c:26257 (10h43m12s: infos 15570/15455 sent/received, bytes 2897447B/3320599B sent/received)
gossip server (0/3 cur/max conns, infos 718643/38636 sent/received, bytes 330616262B/7513830B sent/received)

Note how there's a mixing of `cockroachdb-0` and container IDs, which already seems a bit bad. Also:

ping cockroachdb-0
PING cockroachdb-0 (10.0.0.2): 56 data bytes
92 bytes from b2bb4029cad9 (10.0.0.5): Destination Host Unreachable

That seems pretty bad.

It's also doing this all the time:
I161004 13:22:03.534692 1059834 gossip/client.go:125  [n3] node 3: started gossip client to 63259e48747c:26257
I161004 13:22:03.534902 1059834 gossip/client.go:191  [n3] node 3: sending [node:1 store:3 store:1 first-range store:2 node:2 cluster-id node:3 system-db] to 63259e48747c:26257
I161004 13:22:03.535088 1059834 gossip/client.go:191  [n3] node 3: sending [store:2 node:3 first-range node:1 cluster-id node:2 store:1 store:3 system-db] to 63259e48747c:26257
I161004 13:22:03.535772 1059834 gossip/client.go:132  [n3] node 3: closing client to 63259e48747c:26257: rpc error: code = 2 desc = duplicate connection from node at {tcp d67a7b9fb60c:26257} # d67a7b9fb60c is that container itself, but at least 64259 exists 
```

```
node3 is similarly screwed.

I161004 13:30:33.251354 87 gossip/gossip.go:426  [n3] gossip status (stalled, 3 nodes)
gossip client (1/3 cur/max conns)
  1: cockroachdb-0:26257 (10h51m0s: infos 15643/15636 sent/received, bytes 2905425B/3356026B sent/received)
gossip server (1/3 cur/max conns, infos 724633/31348 sent/received, bytes 333816372B/6279913B sent/received)
  2: b2bb4029cad9:26257 (10h48m29s)
```

**Comment by bdarnell** [comment]
> 1 looks fully connected but is in fact stalled:

Gossip is "stalled" whenever there is no lease holder for the first range, so this is to be expected when the raft transports are broken in this way. 

> Note how there's a mixing of `cockroachdb-0` and container IDs, which already seems a bit bad. 

I've seen a similar mixture of hostnames and IP addresses on the GCE test clusters in the past, although it looks like those are currently using hostnames consistently. Container IDs (and IP addresses) change on restart in this setup, right?

It looks like in this overlay network we have names like `cockroachdb-0` which are resolvable, but `$(hostname)` returns the container ID, which is not what we want. Unless we can configure the container to believe that its hostname is the name in the overlay network rather than the container ID, we'll need to use `--advertise-host` to ensure that we're advertising the hostname that we want other nodes to use.

**Comment by a-robinson** [comment]
> I've seen a similar mixture of hostnames and IP addresses on the GCE test clusters in the past, although it looks like those are currently using hostnames consistently. Container IDs (and IP addresses) change on restart in this setup, right?

Yes, they can change.

> It looks like in this overlay network we have names like cockroachdb-0 which are resolvable, but $(hostname) returns the container ID, which is not what we want. Unless we can configure the container to believe that its hostname is the name in the overlay network rather than the container ID, we'll need to use --advertise-host to ensure that we're advertising the hostname that we want other nodes to use.

We can certainly switch to using `--advertise-host`, but I've been told before that our gossip network should work if a node gets restarted with a new IP/hostname because it'll just come up and start gossiping that new address to the cluster. I take it that's not actually the case?

**Comment by bdarnell** [comment]
Yeah, it should work if the new addresses get gossiped, but this is a scenario that we haven't really tested so I'm not surprised it's failing.

**Comment by a-robinson** [comment]
Ok. @jseldess, for now we should add the `--advertise-host` flag to our Swarm instructions while we separately figure out what's wrong with gossip.

The Kubernetes instructions are fine as is because they're explicitly setting the `--host` flag with a hostname that's consistent across restarts.

**Comment by jseldess** [comment]
I just tried passing `--advertise-host` and had the same problem (sql queries hung). Is there anything wrong with this service creation command? 

```
sudo docker service create --replicas 1 --name cockroachdb-2 --network cockroachdb \
--mount type=volume,source=cockroachdb-2,target=/cockroach/cockroach-data,volume-driver=local \
--stop-grace-period 60s \
cockroachdb/cockroach:beta-20161006 start --advertise-host=cockroachdb-2 --logtostderr --insecure --join=cockroachdb-0:26257
```

**Comment by jseldess** [comment]
Nevermind. @a-robinson pointed out that I'm not running the fixed version of docker. I'll reinstall and try again.

**Comment by a-robinson** [comment]
Yeah, for reference, the cluster was hitting https://github.com/docker/docker/issues/26993

**Comment by bdarnell** [comment]
So is this just a docker bug or is there still something we need to investigate on the cockroach/gossip side?

**Comment by a-robinson** [comment]
There's a real bug with gossip in the clusters that @tschottdorf and I have looked at. @jseldess ran into a docker bug that happens to exhibit the same symptom (SQL CLI commands hanging).

**Comment by jseldess** [comment]
OK, the fixed version of docker is now generally available. I just retested on gce with the new version and with `--advertise-host`, and everything seems to work great. I kill 2 nodes at different times and each started up fine.

**Comment by spencerkimball** [comment]
Closing for lack of follow up. @a-robinson if you believe there is still a real bug in gossip, could you reopen, comment, and assign to me?

**Comment by a-robinson** [comment]
This is still very reproducable if you run the exact the same steps as in the original issue. In other words, follow the instructions in https://www.cockroachlabs.com/docs/orchestrate-cockroachdb-with-docker-swarm-insecure.html but omit all `--advertise-addr` flags. After you've killed one of the processes (and it comes back up with a new hostname but the same data), the cluster will get stuck if you continue trying to use it (which in my case meant running `create database foo` from a sql shell).

**Comment by a-robinson** [comment]
I don't think it's really necessary to fix for 1.0, though, so removing the milestone.

**Comment by a-robinson** [comment]
While the root cause is different, it's interesting to note that this actually gets into a situation a lot like #8832, where the nodes that didn't get restarted keep trying to talk to the other node's previous hostname, so they're never able to open up outgoing connections to the replacement node even though it can talk to them.

I think that if the nodes stopped trying to use the old hostname, things would start working again, but they never do.

**Comment by spencerkimball** [comment]
Is this easy to repro?

**Comment by a-robinson** [comment]
Yes. See my comment above - https://github.com/cockroachdb/cockroach/issues/9658#issuecomment-291612567

**Comment by a-robinson** [comment]
If you think that this actually is 1.0 worthy, I'm happy to get to the bottom of it and fix it.

**Comment by spencerkimball** [comment]
I'd hate to leave this if you've got it repro'ing and have the context
switched in. But balance it with your other priorities. It's not required
for 1.0 because it sounds like a somewhat unusual situation.

**Comment by a-robinson** [comment]
This isn't reproducible on 1.0.2 even though I believe it used to repro every time. The nodes are correctly gossiping their new addresses and updating their node status record when they come up with a new hostname, and the other nodes appear to react correctly to the change. The only bad behavior still around is that one connection per node) reliably gets stuck in grpc for each node that comes up at a new hostname, spamming the logs every second (e.g. `I170706 17:58:44.546692 63 vendor/google.golang.org/grpc/clientconn.go:806  grpc: addrConn.resetTransport failed to create client transport: connection error: desc = "transport: dial tcp: lookup 5eed5d565eca on 127.0.0.11:53: no such host"; Reconnecting to {5eed5d565eca:26257 <nil>}`).

Specifically, there are these two goroutines stuck around on a node that has two of these jammed old connections:

```
goroutine 538 [select]:
github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc.(*addrConn).resetTransport(0xc420b10d00, 0xc420260700, 0x0, 0x2bf3a00)
  /go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/clientconn.go:822 +0x8c1
github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc.(*addrConn).transportMonitor(0xc420b10d00)
  /go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/clientconn.go:902 +0x362
created by github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc.(*ClientConn).resetAddrConn
  /go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/clientconn.go:592 +0x6cb

goroutine 6188 [select]:
github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc.(*addrConn).resetTransport(0xc42033cea0, 0xc4216b2d00, 0x0, 0xc42069a800)
  /go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/clientconn.go:822 +0x8c1
github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc.(*addrConn).transportMonitor(0xc42033cea0)
  /go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/clientconn.go:902 +0x362
created by github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc.(*ClientConn).resetAddrConn
  /go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/clientconn.go:592 +0x6cb
```

That doesn't seem to go away unless you restart the node where the connection is jammed, but that's way less serious than what this issue was originally tracking, so I'll open a separate issue for it.

**Comment by a-robinson** [comment]
And there's no need to open a separate issue, I've sent out https://github.com/cockroachdb/cockroach/pull/16900 instead to reduce the log spam.
